pkg/sign/signer.go

```diff
@@ -15,99 +15,240 @@
 package sign
 
 import (
-	"crypto"
-	"crypto/rand"
-	"crypto/sha256"
-	_ "crypto/sha512" // if user chooses SHA2-384 or SHA2-512 for hash
-	"crypto/x509"
+	"bytes"
 	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
 	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
 
 	protobundle "github.com/sigstore/protobuf-specs/gen/pb-go/bundle/v1"
 	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
 )
 
+const bundleV03MediaType = "application/vnd.dev.sigstore.bundle.v0.3+json"
+
 type Signer interface {
-	Sign(data []byte) (*protobundle.Bundle, error)
+	Sign(content Content, k Keypair) (*protobundle.Bundle, error)
```

Comment:
I think this is the wrong composition. This works fine for the expected PGI workflow, but it creates a binding between the key and CA. If a user needs to sign offline for example, it would not possible.

IMO, Fulcio should be responsible only for requesting a certificate. This is why I prefer the composition of either a higher-level object taking a `KeyPair` and `CA` interface or `Fulcio` struct (You had mentioned a similar composition in another comment), or an interface like what Fredrik proposed.

If we want to avoid doing a larger refactor to the interface, what do you think about something like:

```
type CA interface {
  CertificateRequest(key crypto.PublicKey, identity string, bundle *protobundle.Bundle) (error)
}
```

with the `Fulcio` struct implementing this method.

And in the `main.go` example:

```
// init fulcio opts struct
// init bundle
// get ID token
keypair := sign.NewEphemeralKeyPair()
fulcio := NewFulcio(opts)
fulcio.CertificateRequest(keypair.Public(), idtoken, bundle)
keypair.Sign(content, bundle)
```

Those latter steps can later be composed into a single method later on.

Reply:
I reworked things quite a bit, so individual components like Fulcio or the TimestampAuthority no longer need to understand the bundle format. These changes also include providing the idToken separately when you request a certificate from Fulcio. It did end up simplifying things a bit, and I think it meets the criteria we've laid out, but let me know if there are further changes you want to see.